drivers/pinctrl/pinctrl_wch_afio_20x_30x.c

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright (c) 2024 MASSDRIVER EI (massdriver.space)
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/dt-bindings/pinctrl/ch32v20x_30x-pinctrl.h>
+
+#include <ch32fun.h>
+
+static GPIO_TypeDef *const wch_afio_pinctrl_regs[] = {
+	(GPIO_TypeDef *)DT_REG_ADDR(DT_NODELABEL(gpioa)),
+	(GPIO_TypeDef *)DT_REG_ADDR(DT_NODELABEL(gpiob)),
+	(GPIO_TypeDef *)DT_REG_ADDR(DT_NODELABEL(gpioc)),
+	(GPIO_TypeDef *)DT_REG_ADDR(DT_NODELABEL(gpiod)),
+#if DT_NODE_EXISTS(DT_NODELABEL(gpioe))
+	(GPIO_TypeDef *)DT_REG_ADDR(DT_NODELABEL(gpioe)),
+#endif
+};
+
+int pinctrl_configure_pins(const pinctrl_soc_pin_t *pins, uint8_t pin_cnt, uintptr_t reg)
+{
+	int i;
+
+	for (i = 0; i < pin_cnt; i++, pins++) {
+		uint8_t port = (pins->config >> CH32V20X_V30X_PINCTRL_PORT_SHIFT) & 0x07;
+		uint8_t pin = (pins->config >> CH32V20X_V30X_PINCTRL_PIN_SHIFT) & 0x0F;
+		uint8_t bit0 = (pins->config >> CH32V20X_V30X_PINCTRL_RM_BASE_SHIFT) & 0x1F;
+		uint8_t pcfr_id = (pins->config >> CH32V20X_V30X_PINCTRL_PCFR_ID_SHIFT) & 0x01;
+		uint8_t remap = (pins->config >> CH32V20X_V30X_PINCTRL_RM_SHIFT) & 0x3;
+		GPIO_TypeDef *regs = wch_afio_pinctrl_regs[port];
+		uint32_t pcfr = pcfr_id == 0 ? AFIO->PCFR1 : AFIO->PCFR2;
+		uint8_t cfg = 0;
+
+		if (pins->output_high || pins->output_low) {
+			cfg |= (pins->slew_rate + 1);
+			if (pins->drive_open_drain) {
+				cfg |= BIT(2);
+			}
+			/* Select the alternate function */
+			cfg |= BIT(3);
+		} else {
+			if (pins->bias_pull_up || pins->bias_pull_down) {
+				cfg |= BIT(3);
+			}
+		}
+
+		if (pin < 8) {
+			regs->CFGLR = (regs->CFGLR & ~(0x0F << (pin * 4))) | (cfg << (pin * 4));
+		} else {
+			regs->CFGHR = (regs->CFGHR & ~(0x0F << ((pin - 8) * 4))) |
+				      (cfg << ((pin - 8) * 4));
+		}
+		if (pins->output_high) {
+			regs->OUTDR |= BIT(pin);
+			regs->BSHR |= BIT(pin);
+		} else if (pins->output_low) {
+			regs->OUTDR |= BIT(pin);
+			/* Reset the pin. */
+			regs->BSHR |= BIT(pin + 16);
+		} else {
+			regs->OUTDR &= ~(1 << pin);
+			if (pins->bias_pull_up) {
+				regs->BSHR = BIT(pin);
+			}
+			if (pins->bias_pull_down) {
+				regs->BCR = BIT(pin);
+			}
+		}
+
+		pcfr |= remap << bit0;
+
+		if (pcfr_id == 0) {
+			AFIO->PCFR1 = pcfr;
+		} else {
+			AFIO->PCFR2 = pcfr;
+		}
+
+		if (bit0 == CH32V20X_V30X_PINMUX_USART1_RM) {
```

Comment:
Just checking - is this exceptional case needed? Will there be more in the future, i.e. is it worth reworking the pinmux encoding to cover this rather than special casing it?

Reply:
Yes, it is a exceptional case, the only one for this serie. It is like the one for ch32v003 which additionally has the same for I2C1.